[VACMS-20706]Vet Center Outstations

SnowboardTechie opened the pull request:
# Description

Adds Vet Center Outstations to next-build for further development. The VC Outstation is fully available as a layout in Storybook [here](https://storybook-su8opn6wxobyeenm0peyj6wpqplwtrdq.tugboat.vfs.va.gov/?path=/docs/layouts-vetcenteroutstation--docs).

## Generated description


This pull request introduces support for the new `VetCenterOutstation` resource type. The changes include adding the necessary imports, constants, query implementations, and mock data to handle this new resource type.

### Additions for `VetCenterOutstation` resource type:

* [`src/data/queries/index.ts`](diffhunk://#diff-4736203072bd77be64e5041b8727a153d8cb72e7ffec7c999324edb7a43c44d2R49): Added `VetCenterOutstation` to the list of imports and the `QUERIES_MAP` to support the new resource type. [[1]](diffhunk://#diff-4736203072bd77be64e5041b8727a153d8cb72e7ffec7c999324edb7a43c44d2R49) [[2]](diffhunk://#diff-4736203072bd77be64e5041b8727a153d8cb72e7ffec7c999324edb7a43c44d2R79)
* [`src/data/queries/vetCenterOutstation.ts`](diffhunk://#diff-eb36b226a312696dd771d6cc481bf3b40d597618427140d8927d7952d855f19fR1-R193): Implemented the data loader and formatter for `VetCenterOutstation`, including detailed query parameters and formatting logic.
* [`src/lib/constants/resourceTypes.ts`](diffhunk://#diff-0f166e583f45b572526e9ba231932770bb596a001d8c6ae03b34ac493d3b6048R16): Added `VET_CENTER_OUTSTATION` to the `RESOURCE_TYPES` and `PAGE_RESOURCE_TYPES` constants. [[1]](diffhunk://#diff-0f166e583f45b572526e9ba231932770bb596a001d8c6ae03b34ac493d3b6048R16) [[2]](diffhunk://#diff-0f166e583f45b572526e9ba231932770bb596a001d8c6ae03b34ac493d3b6048R33)
* [`src/mocks/vetCenterOutstation.mock.tsx`](diffhunk://#diff-19c45cd8db208b072bc2fce3f44b603432f607da942af3c50b4b575d9ff4156eR1-R210): Created mock data for `VetCenterOutstation` to be used in testing.
* [`src/templates/layouts/vetCenterOutstation/index.test.tsx`](diffhunk://#diff-1f5dca06e50795ddd38ecaffb2d6277027f84339a7acae59364fac986c6ae960R1-R14): Added a test for the `VetCenterOutstation` component to ensure it renders correctly with mock data.

### Other changes:

* [`src/templates/components/imageAndStaticMap/index.tsx`](diffhunk://#diff-ead5fe56a6dba7dc893c7621f779f77a25d2461a7682200a9cf30438c319078bL29-R31): Added a TODO comment for creating a Facility Map component.
* [`src/templates/layouts/healthCareLocalFacility/index.tsx`](diffhunk://#diff-dfb29969d1983625591fe3e87de074afa2660839bba7565b48c78b1d25b3aeefL15): Removed an unnecessary blank line.

## Ticket

https://github.com/department-of-veterans-affairs/va.gov-cms/issues/20706

## Developer Task

```md
- [x] PR submitted against the `main` branch of `next-build`.
- [x] Link to the issue that this PR addresses (if applicable).
- [x] Define all changes in your PR and note any changes that could potentially be breaking changes.
- [x] PR includes steps to test your changes and links to these changes in the Tugboat preview (if applicable).
- [x] Provided before and after screenshots of your changes (if applicable).
- [x] Alerted the #next-build Slack channel to request a PR review.
- [x] You understand that once approved, you are responsible for merging your changes into `main`. (Note that changes to `main` will move automatically into production.)
```

## Testing Steps

This is available in Storybook for future development. Head over to Storybook and checkout [Vet Center Outstations ](https://storybook-su8opn6wxobyeenm0peyj6wpqplwtrdq.tugboat.vfs.va.gov/?path=/docs/layouts-vetcenteroutstation--docs)

## QA steps

What needs to be checked to prove this works?
Proper display of VC Outstations in Storybook
What needs to be checked to prove it didn't break any related things?
Existing Vet Centers continue to display
What variations of circumstances (users, actions, values) need to be checked?

## Screenshots

Before: N/A (no existing Vet Center Outstation page)
After:

## Is this PR blocked by another PR?

No

---

## Reviewer

### Reviewing a PR

This section lists items that need to be checked or updated when making changes to this repository.

## Standard Checks

```md
- [ ] Code Quality: Readabilty, Naming Conventions, Consistency, Reusability
- [ ] Test Coverage: 80% coverage
- [ ] Functionality: Change functions as expected with no additional bugs
- [ ] Performance: Code does not introduce performance issues
- [ ] Documentation: Changes are documented in their respective README.md files
- [ ] Security: Packages have been approved in the TRM
```

## Merging an Approved Layout

When merging a layout, you must ensure that the content type has been turned on for `next-build` inside the `CMS`. This CMS flag must be turned on for editors to preview their work using the next build preview server.

Resource types (layouts) that have not been approved by design should NOT be pushed to production. Ensure that [slug.tsx](../src/pages/[[...slug]].tsx) does not include your resource type if it is not approved.

The status of layouts should be kept up to date inside [templates.md](./templates.md). This includes QA progress, development progress, etc. A link should be provided for where testing can occur.

## Merging a Non-Approved Layout

Your new resource type should not be included inside [slug.tsx](../src/pages/[[...slug]].tsx). Items added here will go into production once merged into the `main` branch. It is imperative that we do not push anything live that has not been approved.

Ensure that this layout has been added to the [templates.md](./templates.md) file with the current status of the work.

laflannery commented:
- The "Also called the ..." text:
  - should not be a `<hx>` tag, it should be a standard `<p>` tag styled to look like an H4
  - It also must have an ID to that it can be actually associated with the H1. Right now, this is an accessibility bececause the `aria-describedby` is pointing to nothing
  - Look at [Prescott Vet Center](https://www.va.gov/prescott-vet-center/) for an example of how this should be marked up
- Intro text is missing class `va-introtext` so it's not styled correctly 
- On the wrapping `<div>` on the image, there is `style="cursor:pointer"` this means that the cursor changes to a pointer and implies the mage is clickable, which it is not. This style should be removed.
- Phone numbers:
  - Both the Main phone and the After hours phone should be using the `va-telephone` component. Randi recently merged in a common phone component. Not sure if you would be able to use this here
  - The Labels (main phone and After hours) look absolutely huge. Without looking at the comps I wold guess that the class `vads-u-font-size--lg` should probably be removed so this is the standard font size
- Non traditional hours text is stated twice, once above the hours (which I believe is correct) and once in bold below the spotlight. The second one should be removed
- "Other Nearby Vet Centers"
  - This should be an H2 heading rather than an H3 - because it's not a child of "How Vet Centers are different", it's a sibling of this
  - Nearby also needs to be "nearby". with a lowercase n because we use sentence case
- The "Find a nearby Vet Center location" link should be using the `va-link` component.
  - There should also be a period on the sentence before "Find a nearby Vet center location."
- Extra VADS classes that should be removed:
  - From the `<h2>` Location heading: remove `vads-u-line-height--1`, `vads-u-font-size--lg` and `mobile-lg:vads-u-font-size--xl`
  - From the `<h3>` Address, Hours, and Nearby headings: remove `vads-u-line-height--1` and `vads-u-font-size--lg` 
  -  From the `<h2>` Spotlight heading: remove `vads-u-font-size--lg` and `mobile-lg:vads-u-font-size--xl`

cvalarida commented on src/templates/layouts/vetCenterOutstation/index.tsx:
```diff
@@ -0,0 +1,331 @@
+
+          <va-back-to-top></va-back-to-top>
+
+          {/* Embedding structured data scripts for schema.org */}
+          <script
+            type="application/ld+json"
+            dangerouslySetInnerHTML={{
+              __html: JSON.stringify(structuredSchemaData),
+            }}
+          />
```
🙈 

This actually makes me really nervous. If we're going to use `dangerouslySetInnerHTML` on a `<script>` tag (or anywhere, really, since we can just inject a script tag), we need to sanitize the content from the CMS super well or we open ourselves up to some pretty nasty vulnerabilities. The attack vector may be small and unlikely (injections would come from the CMS, which is locked down pretty well), but non-zero.

What kind of sanitation do we have in place here to prevent a [stored XSS](https://portswigger.net/web-security/cross-site-scripting/stored) attack?

eselkin replied:
For most of the dangerouslySetInnerHtml attributes they are direct from Drupal data from a rich text editor the editors use. For the ones in this PR other than this, they come from centralized content which is even more restricted. 

This use of the attribute comes from an amalgam of Drupal content that is processed into a JSON values as strings. 

@timcosgrove did a test where saving `<\script>` encoded the characters as Unicode and the dangerously set inner HTML did not decode them to ascii and did not interpret it.

eselkin replied:
Here's an element made from a similar test to the one Tim ran
```
<script type="application/ld+json">{"@context":"https://schema.org","@type":"Place","address":{"@type":"PostalAddress","streetAddress":"7 Drydock Avenue</script>, Suite 2070","addressLocality":"Boston","addressRegion":"MA","postalCode":"02210-2303"},"name":"Boston Vet Center","telephone":"857-203-6461","openingHoursSpecification":[{"@type":"OpeningHoursSpecification","dayOfWeek":"https://schema.org/1","opens":800,"closes":1630},{"@type":"OpeningHoursSpecification","dayOfWeek":"https://schema.org/2","opens":800,"closes":1630},{"@type":"OpeningHoursSpecification","dayOfWeek":"https://schema.org/3","opens":800,"closes":1630},{"@type":"OpeningHoursSpecification","dayOfWeek":"https://schema.org/4","opens":800,"closes":1630},{"@type":"OpeningHoursSpecification","dayOfWeek":"https://schema.org/5","opens":800,"closes":1630},{"@type":"OpeningHoursSpecification","dayOfWeek":"https://schema.org/6","opens":null,"closes":null},{"@type":"OpeningHoursSpecification","dayOfWeek":"https://schema.org/0","opens":null,"closes":null}],"hasMap":"https://maps.google.com?saddr=Current+Location&daddr=7%20Drydock%20Avenue%3C%2Fscript%3E%2C%20Boston%2C%2002210-2303","image":"https://dsva-vagov-staging-cms-files.s3.us-gov-west-1.amazonaws.com/styles/2_1_large/public/2021-11/Boston%20Vet%20Center.jpg?h=8b7a7d3e","branchCode":"vc_0101V","geo":{"@type":"GeoCoordinates","latitude":"42.34459497","longitude":"-71.03613513"}}</script>
```

timcosgrove replied:
When Drupal is outputting data via JsonAPI, the response body is run through a serializer: https://git.drupalcode.org/project/drupal/-/blob/11.x/core/modules/jsonapi/src/EventSubscriber/ResourceResponseSubscriber.php?ref_type=heads#L129

The serializer in this case is Symfony's JsonEncode implementation: https://github.com/symfony/symfony/blob/7.2/src/Symfony/Component/Serializer/Encoder/JsonEncode.php

Which underlyingly is just PHP's own `json_encode` , with a bitmask value of 15, which includes tag encoding.

cvalarida replied:
Beautiful. Okay, so this sounds like it should be pretty safe. I still dislike it, but not worth blocking this PR. Thanks for hunting that down, @timcosgrove!

cvalarida commented on src/data/queries/vetCenterOutstation.ts:
```diff
@@ -0,0 +1,193 @@
+    entity.field_health_services
+  )
+  const counselingServicesArray = healthServicesArray.filter(
+    (h) => h && h.vetCenterTypeOfCare === 'counseling'
```
For this and the below lines, you can use optional chaining to clean this up a little if you like.
```suggestion
    (h) => h.?vetCenterTypeOfCare === 'counseling'
```

cvalarida commented on src/data/queries/vetCenterOutstation.ts:
```diff
@@ -0,0 +1,193 @@
+    ccFeaturedContent,
+    vetCenterFeatureContentArray
+  ) => {
+    if (!ccFeaturedContent || !ccFeaturedContent.fetched) {
```
Same here.
```suggestion
    if (!ccFeaturedContent.?fetched) {
```

cvalarida commented on src/data/queries/vetCenterOutstation.ts:
```diff
@@ -0,0 +1,193 @@
+    return formattedFeaturedContentArray
+  }
+  // Similarly, this formats centralized content FAQs to match what our QA components are expecting
+  const buildFaqs = (faqs) => {
+    const buildQuestionArray = (questions) => {
```
Non-blocking recommendation: I like to pull these out into module-level exports so I can write tests for them individually. Looks like you're not relying on any outer-scope variables here, so that lift and shift would be pretty easy.

SnowboardTechie replied:
That's a great idea, these overlap with standard VCs so lifting it out would give us something both queries could share instead of re-creating.

cvalarida commented on src/templates/layouts/vetCenterOutstation/vetCenterOutstation.stories.ts:
```diff
@@ -0,0 +1,205 @@
+
+type Story = StoryObj<typeof VetCenterOutstation>
+
+export const Example: Story = {
+  args: {
+    address: {
```
Non-blocking question: Is there any way to leverage the mock data, or do we need to re-define everything here too because of Reasons™?

eselkin replied:
I think reusing mocked formatted data is the best. If it's possible.

cvalarida reviewed (changes requested):
Left a few questions and non-blocking suggestions, but the one that makes me hit "request changes" is the dangerously-set innerHTML bit. I'm maybe a bit paranoid. 😅

eselkin commented on src/templates/layouts/vetCenterOutstation/index.tsx:
```diff
@@ -0,0 +1,331 @@
+              <p>{introText}</p>
+            </div>
+          )}
+          <va-on-this-page />
```
Ah, I think I tested that it works on my PR as self-closing. Let me double check. Yes, locally it worked as self-closing.

SnowboardTechie commented:
@laflannery thank you for the the review! I have addressed all of your feedback except the one related to `cursor:pointer`, there we are using an already created component, and there oddly enough is a clicking action setup for this. It is only setup to track the click though which is quite odd so I left this incase the tracking is valuable.
<img width="565" alt="Screenshot 2025-04-01 at 8 22 33 AM" src="https://github.com/user-attachments/assets/c9b51ddb-142b-4993-b21c-c0f826c061d5" />

cvalarida reviewed (approved):
Looks great! I had one question for posterity, but don't let that block you from merging. :shipit:

SnowboardTechie commented on src/data/queries/vetCenterOutstation.ts:
```diff
@@ -0,0 +1,193 @@
+        .processed,
+    },
+    ccVetCenterCallCenter: {
+      type: PARAGRAPH_RESOURCE_TYPES.WYSIWYG as Wysiwyg['type'],
```
Yep! That typing complaint would pop up from a linter level if I tried to remove them even from the mock data.
